packages/vincent-contracts/src/facets/VincentAppFacet.sol

```
@@ -0,0 +1,614 @@
+
+                // Validate non-empty policy IPFS CID
+                if (bytes(policyIpfsCid).length == 0) {
+                    revert EmptyPolicyIpfsCidNotAllowed(appId, i);
+                }
+
+                bytes32 hashedToolPolicy = keccak256(abi.encodePacked(policyIpfsCid));
+
+                // Step 6.1: Add the policy hash to the ToolPolicies
+                toolPoliciesStorage.policyIpfsCidHashes.add(hashedToolPolicy);
+
+                // Step 6.2: Store the policy IPFS CID globally if it's not already stored.
+                if (bytes(ts.ipfsCidHashToIpfsCid[hashedToolPolicy]).length == 0) {
+                    ts.ipfsCidHashToIpfsCid[hashedToolPolicy] = policyIpfsCid;
+                }
+
+                // Create a new Policy storage structure for this policy in the current app version
+                VincentAppStorage.Policy storage policy =
+                    toolPoliciesStorage.policyIpfsCidHashToPolicy[hashedToolPolicy];
+
+                // Step 7: Get the Policy parameter name hashes for this policy
+                EnumerableSet.Bytes32Set storage policyParameterNameHashes = policy.policyParameterNameHashes;
+
+                // Step 8: Iterate through policy parameters.
+                uint256 paramCount = toolPolicyParameterNames[i][j].length;
+
+                for (uint256 k = 0; k < paramCount; k++) {
+                    string memory paramName = toolPolicyParameterNames[i][j][k]; // Cache calldata value
+
+                    // Check for empty parameter name
+                    if (bytes(paramName).length == 0) {
+                        revert EmptyParameterNameNotAllowed(appId, i, j, k);
+                    }
+
+                    bytes32 hashedPolicyParameterName = keccak256(abi.encodePacked(paramName));
+
+                    // Step 8.1: Register the policy parameter.
+                    policyParameterNameHashes.add(hashedPolicyParameterName);
```

Check for duplicate policyParamName. We should throw an error for: ["maxSpend", "maxSpend", "period"]. Also add a test for it

---

Addressed by [this commit](https://github.com/LIT-Protocol/Vincent/pull/11/commits/4d1599f2ef270efbbbced321a7235c8d4ee36ab6)